examples/gno.land/p/demo/uuid/uuid.gno

```diff
@@ -0,0 +1,115 @@
+// Copyright 2023 The godruoyi/gno-snowflake Author. All rights reserved.
+// Use of this source code is governed by a MIT.
+
+package uuid
+
+import (
+	"errors"
+	"time"
+)
+
+// These constants are the bit lengths of snowflake ID parts.
+const (
+	TimestampLength uint8  = 41
+	MachineIDLength uint8  = 10
+	SequenceLength  uint8  = 12
+	MaxSequence     uint16 = 1<<SequenceLength - 1
+	MaxTimestamp    uint64 = 1<<TimestampLength - 1
+	MaxMachineID    uint64 = 1<<MachineIDLength - 1
+
+	machineIDMoveLength = SequenceLength
+	timestampMoveLength = MachineIDLength + SequenceLength
+)
+
+// default start time is 2008-11-10 23:00:00 UTC, why ? In the playground the time begins at 2009-11-10 23:00:00 UTC.
+// It can run on golang playground.
+// default machineID is 0
+type SequenceResolver func(ms int64) (uint16, error)
+
+var (
+	resolver      SequenceResolver
+	MachineID     uint64 = 0
+	sequence      uint16 = 0
+	StartTime            = time.Date(2008, 11, 10, 23, 0, 0, 0, time.UTC)
+	LastTimestamp int64  = 0
+	// _             int    //will not sync atomic Gno not goroutine, implement in the future for the resolver.
+)
+
+// ID use ID to generate snowflake id, and it will ignore error. if you want error info, you need use NextID method.
+func ID() uint64 {
+	id, _ := NextID()
+	return id
+}
+
+// NextID use NextID to generate snowflake id and return an error.
+func NextID() (uint64, error) {
+	current := currentMillis()
+	seqResolver := callSequenceResolver()
+	seq, err := seqResolver(current)
+	if err != nil {
+		return 0, err
+	}
+
+	for seq >= MaxSequence {
+		current = waitForNextMillis(current)
+		seq, err = seqResolver(current)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	if current < LastTimestamp {
+		return 0, errors.New("Invalid System Clock")
+	}
+
+	LastTimestamp = current
+	elapsedTime := current - StartTime.UTC().UnixNano()/1e6
+	if elapsedTime < 0 || uint64(elapsedTime) > MaxTimestamp {
+		return 0, errors.New("Exceeding the time limit")
+	}
+
+	id := (uint64(elapsedTime) << timestampMoveLength) |
+		(uint64(MachineID) << machineIDMoveLength) |
+		uint64(seq)
+	return id, nil
+}
+
+// SetMachineID specify the machine ID. It will panic when machined > max limit for 2^10-1.
+func SetMachineID(m uint64) {
+	if m > MaxMachineID {
+		panic("The machineID cannot be greater than 1023")
+	}
+	MachineID = m
+}
+
+// SetSequenceResolver define the sequence resolver.
+func SetSequenceResolver(seq SequenceResolver) {
+	if seq != nil {
+		resolver = seq
+	}
+}
+
+func waitForNextMillis(last int64) int64 {
```

Comment:
What is this function meant to do? If `time.Now()` is called multiple times within the same transaction, it will always return the same value, so it seems like this will just always return the current time.  It's also worth noting that `time.Now()` only returns time with a granularity of seconds, so using `UnixNano()` over `Unix()` has no real effect.

Reply:
To avoid collision when generating `UUID`, this function allows waiting until the current time exceeds the previous time to obtain a unique `timestamp`. This prevents two `UUIDs` from having the same `timestamp`. My use of `UnixNano()`, as in the [Snowflake](https://github.com/godruoyi/go-snowflake/blob/v0.0.2/snowflake.go) package, comes from the fact that I saw it was created in the [time.gno](https://github.com/gnolang/gno/blob/master/gnovm/stdlibs/time/time.gno) package

Reply:
Try writing a test where this block of code gets hit in `NextID`:
```go
for seq >= MaxSequence {
	current = waitForNextMillis(current)
	seq, err = seqResolver(current)
	if err != nil {
		return 0, err
	}
}
```
Or try testing `waitForNextMillis` by calling it directly. When I tried this I found that it entered an infinite loop. This is because this function is called from one place and it passes in the value obtained by calling `currentMillis`, or `time.Now`. Because this value returned will always be the same for calls within the same transaction, it will block any further execution.

Can you confirm you are seeing the same result?

Reply:
I just tested it as well and I get the same result as you. I understand better what you explained to me, I hadn’t thought to test that, and since UnixNano() is present in time.gno, I didn’t think it would cause an infinite loop. Thank you for the explanation, I will correct this and be more careful 👍🏼

Reply:
To be clear, it is not `UnixNano` that is causing an infinite loop. If is that `last` has the value of `time.Now()` and `currentMillis()` also returns the value of `time.Now()`, so in this case `now` and `last` will always be equal so `for now <= last` will always evaluate to true.  Sorry if I'm repeating what you already now; I just want to make sure we have the same understanding.

Reply:
Yes, we have the same understanding, sorry it was me who wasn’t precise enough in the explanation.

Reply:
The "waitForNext" functionality should be removed. Instead, your algorithm should rely on a single static timestamp and increment a local counter in your struct to simulate the increasing milliseconds.

Reply:
[`2f1bc74`](https://github.com/gnolang/gno/pull/2076/commits/2f1bc74778a2dd8937f8bcbe5c51ee087edbf249)